File: src/indexer/languages/ruby.ts
import type { LanguageParser } from "./parser.js";
import {
  buildTreeSitter,
  field,
  named,
  type Node,
  type Emit,
  type Ctx,
} from "./treesitter/base.js";

function emitMethod(node: Node, emit: Emit, container?: string): void {
  const nameNode = field(node, "name");
  if (!nameNode) return;
  if (container) {
    emit.symbol({ name: `${container}.${nameNode.text}`, kind: "method", node, nameNode, exported: false, simpleName: nameNode.text });
  } else {
    emit.symbol({ name: nameNode.text, kind: "function", node, nameNode, exported: true });
  }
}

function processBody(body: Node, emit: Emit, ctx: Ctx, container?: string): void {
  for (const node of body.namedChildren) {
    switch (node.type) {
      case "method":
      case "singleton_method":
        emitMethod(node, emit, container);
        break;
      case "class":
      case "module": {
        const nameNode = field(node, "name");
        if (!nameNode) break;
        emit.symbol({ name: nameNode.text, kind: "class", node, nameNode, exported: true });
        const inner = field(node, "body");
        if (inner) processBody(inner, emit, ctx, nameNode.text);
        break;
      }
      case "assignment": {
        const left = field(node, "left");
        if (left && left.type === "constant") {
          emit.symbol({ name: left.text, kind: "const", node: left, nameNode: left, exported: true, signature: left.text });
        }
        break;
      }
      case "call": {
        const method = field(node, "method");
        if (method && (method.text === "require" || method.text === "require_relative")) {
          const args = field(node, "arguments");
          const str = args ? named(args, "string") : undefined;
          const raw = str ? str.text.replace(/^['"]|['"]$/g, "") : "";
          if (raw) {
            const to = method.text === "require_relative"
              ? (findSuffix(ctx.relSet, `${raw}.rb`) ?? `${raw}.rb`)
              : raw;
            emit.import({ from: ctx.file, to, names: [] });
          }
        }
        break;
      }
    }
  }
}

function findSuffix(relSet: Set<string>, rel: string): string | null {
  if (relSet.has(rel)) return rel;
  for (const r of relSet) if (r.endsWith(`/${rel}`)) return r;
  return null;
}

function extract(root: Node, emit: Emit, ctx: Ctx): void {
  processBody(root, emit, ctx);
}

export const rubyParser: LanguageParser = {
  name: "ruby",
  extensions: ["rb"],
  build: (root, files) => buildTreeSitter(root, files, "ruby", extract),
};

File: rust/sens-hook/src/lang/ruby.rs
use std::collections::HashSet;

use tree_sitter::Node;

use super::treesitter::{
    declare, field, named, never_qualified, text, Emitted, Extra, Options, Scope,
};

fn emit_method(node: &Node, source: &str, container: Option<&str>, out: &mut Emitted) {
    let Some(name_node) = field(node, "name") else { return };
    let simple = text(&name_node, source).to_string();
    match container {
        Some(c) => declare(out, source, format!("{c}.{simple}"), "method", node, &name_node, false, Extra { simple_name: Some(simple), ..Extra::default() }),
        None => declare(out, source, simple, "function", node, &name_node, true, Extra::default()),
    }
}

fn find_suffix(rel_set: &HashSet<String>, rel: &str) -> Option<String> {
    if rel_set.contains(rel) {
        return Some(rel.to_string());
    }
    rel_set.iter().find(|r| r.ends_with(&format!("/{rel}"))).cloned()
}

fn process_body(
    body: &Node,
    source: &str,
    file: &str,
    rel_set: &HashSet<String>,
    container: Option<&str>,
    out: &mut Emitted,
) {
    let mut cursor = body.walk();
    let children: Vec<Node> = body.named_children(&mut cursor).collect();
    for node in children {
        match node.kind() {
            "method" | "singleton_method" => emit_method(&node, source, container, out),
            "class" | "module" => {
                let Some(name_node) = field(&node, "name") else { continue };
                let cname = text(&name_node, source).to_string();
                declare(out, source, cname.clone(), "class", &node, &name_node, true, Extra::default());
                if let Some(inner) = field(&node, "body") {
                    process_body(&inner, source, file, rel_set, Some(&cname), out);
                }
            }
            "assignment" => {
                if let Some(left) = field(&node, "left") {
                    if left.kind() == "constant" {
                        let name = text(&left, source).to_string();
                        let sig = Some(name.clone());
                        declare(out, source, name, "const", &left, &left, true, Extra { signature: sig, ..Extra::default() });
                    }
                }
            }
            "call" => {
                let Some(method) = field(&node, "method") else { continue };
                let method_name = text(&method, source);
                if method_name != "require" && method_name != "require_relative" {
                    continue;
                }
                let raw = field(&node, "arguments")
                    .and_then(|args| named(&args, "string"))
                    .map(|s| text(&s, source).trim_matches(|c| c == '\'' || c == '"').to_string())
                    .unwrap_or_default();
                if raw.is_empty() {
                    continue;
                }
                let to = if method_name == "require_relative" {
                    let candidate = format!("{raw}.rb");
                    find_suffix(rel_set, &candidate).unwrap_or(candidate)
                } else {
                    raw
                };
                out.import(file, to, Vec::new());
            }
            _ => {}
        }
    }
}

pub fn extract(root: &Node, source: &str, file: &str, rel_set: &HashSet<String>, out: &mut Emitted) {
    process_body(root, source, file, rel_set, None, out);
}

pub fn options() -> Options {
    Options { scope: Scope::Name, qualified_use: never_qualified }
}
